src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+        }
+        try {
+            return Timestamp.createFromUtcFields(
+                    precision,
+                    year,
+                    month,
+                    day,
+                    hour,
+                    minute,
+                    second,
+                    fractionalSecond,
+                    offset
+            );
+        } catch (IllegalArgumentException e) {
+            throw new IonException("Illegal timestamp encoding. ", e);
+        }
+    }
+
+    /**
+     * Reads the annotation symbol IDs from an annotation wrapper.
+     * @param sidConsumer the sink for the symbol IDs. Note: in newer Java versions, this could be a Consumer, reducing
+     *                    the number of loops required to, e.g., convert symbol IDs into SymbolTokens.
+     */
+    private void readAnnotationSids(List<Integer> sidConsumer) {
+        int savedPeekIndex = peekIndex;
+        peekIndex = annotationStartPosition;
+        long annotationsEndPosition = peekIndex + annotationsLength;
+        while (peekIndex < annotationsEndPosition) {
+            sidConsumer.add(readVarUInt());
+        }
+        peekIndex = savedPeekIndex;
+    }
+
+    /**
+     * Gets the annotation symbol IDs for the current value, reading them from the buffer first if necessary.
+     * @return the annotation symbol IDs, or an empty list if the current value is not annotated.
+     */
+    private List<Integer> getAnnotationSids() {
+        if (containerStack.isEmpty()) {
+            return lookahead.getAnnotationSids();
+        } else {
+            if (annotationSids.isEmpty()) {
+                readAnnotationSids(annotationSids);
+            }
+            return annotationSids;
+        }
+    }
+
+    @Override
+    public String[] getTypeAnnotations() {
+        if (hasAnnotations) {
+            List<Integer> annotationSids = getAnnotationSids();
+            String[] annotationArray = new String[annotationSids.size()];
+            for (int i = 0; i < annotationArray.length; i++) {
+                String symbol = getSymbol(annotationSids.get(i));
+                if (symbol == null) {
+                    throw new UnknownSymbolException(annotationSids.get(i));
+                }
+                annotationArray[i] = symbol;
+            }
+            return annotationArray;
+        }
+        return _Private_Utils.EMPTY_STRING_ARRAY;
+    }
+
+    @Override
+    public SymbolToken[] getTypeAnnotationSymbols() {
+        if (hasAnnotations) {
+            List<Integer> annotationSids = getAnnotationSids();
+            SymbolToken[] annotationArray = new SymbolToken[annotationSids.size()];
+            for (int i = 0; i < annotationArray.length; i++) {
+                annotationArray[i] = getSymbolToken(annotationSids.get(i));
+            }
+            return annotationArray;
+        }
+        return SymbolToken.EMPTY_ARRAY;
+    }
+
+    private static final Iterator<String> EMPTY_ITERATOR = new Iterator<String>() {
```

Comment:
I was surprised you weren't using [`Collections.emptyIterator()`](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#emptyIterator()) and then I remembered: Java 5. 😐